### ISSUE TITLE
V4 does not await async functions in Node.js properly.

### ISSUE THREAD

**Issue by Ayzrian** [Azure/azure-functions-nodejs-worker#652]
Tested the same code on v3 and v4. Works on v3, but when we run it on v4 looks like the runtime does not awaits the async code execution properly.

**Comment by ejizba** [comment]
Hi @Ayzrian I moved this to the Node.js-specific repo. Can you give a little more information: for example what specific version of v4 are you on? And could you post either logs from running the function or some sample code demonstrating the issue?

**Comment by Ayzrian** [comment]
Hey, @ejizba I have tracked down the issue. It is related to https://github.com/microsoft/ApplicationInsights-node.js/blob/c59867afb06625116854f492078e8148a8348d08/AutoCollection/AzureFunctionsHook.ts this logic was introduced in v2.4.0 three days ago when I started to see this issue.

The root cause is in this PR https://github.com/microsoft/ApplicationInsights-node.js/pull/1044 

File AzureFunctionsHook.ts, they call originalCallback on line 52 but do not handle a case when the callback could have returned a promise. That lead to problem of function core tools thinking that the execution finished while the actually function code is still running. 

@hectorhdzg does it makes sense to you? If yes, can we fix that, please? I can reproduce the issue when using 2.4.0 version of app insights package, but if I downgrade all works.

**Comment by Ayzrian** [comment]
Here is the minimal code to reproduce the issue.


```js

const appInsights = require('applicationinsights');

appInsights.setup().start();

module.exports = async function (context, req) {
    context.log('JavaScript HTTP trigger function processed a request.');

    await new Promise((res) => setTimeout(() => {
        const name = (req.query.name || (req.body && req.body.name));
        const responseMessage = name
          ? "Hello, " + name + ". This HTTP triggered function executed successfully."
          : "This HTTP triggered function executed successfully. Pass a name in the query string or in the request body for a personalized response.";

        context.res = {
            // status: 200, /* Defaults to 200 */
            body: responseMessage
        };

        res();
    }, 1000))

    context.log.info('Finished function execution.')
}

```

Run it with applicationinsights@2.4.0 and you will see something like this in the logs


[2023-01-14T19:40:15.868Z] Executing 'Functions.HttpTrigger' (Reason='This function was programmatically called via the host APIs.', Id=84b1950d-84c4-487a-81ac-632cad8c3e00)
[2023-01-14T19:40:15.908Z] JavaScript HTTP trigger function processed a request.
[2023-01-14T19:40:15.943Z] Executed 'Functions.HttpTrigger' (Succeeded, Id=84b1950d-84c4-487a-81ac-632cad8c3e00, Duration=87ms)
[2023-01-14T19:40:16.800Z] Executing 'Functions.HttpTrigger' (Reason='This function was programmatically called via the host APIs.', Id=f9c282bc-0e0e-477d-ad33-3a1ed11357b0)
[2023-01-14T19:40:16.805Z] JavaScript HTTP trigger function processed a request.
[2023-01-14T19:40:16.805Z] Executed 'Functions.HttpTrigger' (Succeeded, Id=f9c282bc-0e0e-477d-ad33-3a1ed11357b0, Duration=6ms)
[2023-01-14T19:40:16.904Z] Warning: Unexpected call to 'log' on the context object after function execution has completed. Please check for asynchronous calls that are not awaited or calls to 'done' made before function execution completes. Function name: HttpTrigger. Invocation Id: 84b1950d-84c4-487a-81ac-632cad8c3e00. Learn more: https://go.microsoft.com/fwlink/?linkid=2097909
[2023-01-14T19:40:17.805Z] Warning: Unexpected call to 'log' on the context object after function execution has completed. Please check for asynchronous calls that are not awaited or calls to 'done' made before function execution completes. Function name: HttpTrigger. Invocation Id: f9c282bc-0e0e-477d-ad33-3a1ed11357b0. Learn more: https://go.microsoft.com/fwlink/?linkid=2097909



While with 2.3.6 for example it looks good.

[2023-01-14T19:41:37.571Z] Executing 'Functions.HttpTrigger' (Reason='This function was programmatically called via the host APIs.', Id=cf11f23e-7f0c-4eee-a8b7-82def2c18dfd)
[2023-01-14T19:41:37.605Z] JavaScript HTTP trigger function processed a request.
[2023-01-14T19:41:38.608Z] Finished function execution.
[2023-01-14T19:41:38.664Z] Executed 'Functions.HttpTrigger' (Succeeded, Id=cf11f23e-7f0c-4eee-a8b7-82def2c18dfd, Duration=1104ms)

**Comment by chuanqisun** [comment]
(this comment is moved to upstream repo: https://github.com/microsoft/ApplicationInsights-node.js/issues/1063)

**Comment by ejizba** [comment]
Hi @chuanqisun we don't have much more to say on this issue, @Ayzrian pretty much said everything already! 

The issue is in v2.4.0 of the applicationinsights package and the only mitigation at the moment is to downgrade to v2.3.6. I recommend you follow this issue for further updates: https://github.com/microsoft/ApplicationInsights-node.js/issues/1063 from the app insights team

**Comment by chuanqisun** [comment]
sorry @ejizba, I was confused by which repo contains the actual root cause code. I should have noticed your `external` tag. I will post an update there. Thanks!

**Comment by ejizba** [comment]
Closing as the applicationinsights package released v2.4.1 a few days ago to address this bug. Thanks for reporting and all the details @Ayzrian!
